### ISSUE TITLE
FIX for Cumulative Layout Shift (CLS) issue with Pro active chat invitations

### ISSUE THREAD

**Issue by Queue-Fair** [LiveHelperChat/livehelperchat#1879]
Hi there, we love LHC!  Thank you!!!!

Google now penalises sites in search rankings if they have a high Cumulative Layout Shift.   After much debugging, we determined that the source of the CLS on our site was the appearance of the pro-active chat invitation that we show after a few seconds on each page.

This is because when the invitation iframe is first shown, it is centred in the screen regardless of its final position, then its style is set to display: block !important, and then it is moved to its final position (in the bottom right of the window in our case).  It is the final move that causes the high CLS (greater than 0.1), because its display is already "block".

We couldn't figure out how to change the order in which this happens with the index.js logic, but we did manage to kludge-fix it by modifying design/defaultheme/js/widgetv2/index.js , which is minified so all the code is one line.

First back up this file, then find the first "show" function on the line:

{key:"show",value:function(){this.restyle("display","block !important")}}

Change to look like this:

{key:"show",value:function(){setTimeout(()=>{this.restyle("display", "block !important")}, 100)}}

The timeout of 100ms allows the iframe for the pro-active chat invitation to be moved into position before its display is set to block, thus eliminating layout shift.

This fixed the CLS but we also had to add a style for the iframe to get it to display in the right place:
<style>
#lhc_widget_v2 {
        height: auto !important;
        max-height: 145px !important;
}
</style>

I hope this helps someone!  We would have had to abandon LiveHelperChat if we hadn't figured this out, as Google search rankings are so important.

**Comment by remdex** [comment]
Thanks for feed, I'll try to investigate and put fix in official version. 

As for most website I use LHC I have nginx there and request to lhc is blocked via nginx and HTTP server level.

https://doc.livehelperchat.com/docs/nginx-configuration-tips

**Comment by remdex** [comment]
After fix and changes
![image](https://user-images.githubusercontent.com/1146085/211504925-8ca8d169-e7ae-403c-94b5-996228cc6222.png)

But as I mentioned for nginx I just block all google request, so lhc does not influence rank at all.

**Comment by Queue-Fair** [comment]
Ah the problem with Lighthouse is it only picks up CLS during page load.  The CLS due to the pro active invite happens after that (we don't show an invite for at least 10 seconds) - but even though it doesn't show in lighthouse, Google still uses the CLS after the page load for search rankings - so if you only use Lighthouse to test this you might think it's fixed but still have a problem.

So, you might find the following code snippet helpful for checking CLS throughout the lifetime of the page:

<script>
let cls = 0;
new PerformanceObserver((entryList) => {
  for (const entry of entryList.getEntries()) {
    if (!entry.hadRecentInput) {
      cls += entry.value;
      console.log('Current CLS value:', cls, entry);
    }
  }
}).observe({type: 'layout-shift', buffered: true});
</script>

The console will log each specific element that caused layout shift, where it was moved from and where it was moved to.

Hope that's helpful!

**Comment by remdex** [comment]
Thanks, I'll see tomorrow if I can improve it that way also.

**Comment by remdex** [comment]
Ok, so even after your code snippet. I don't see this issue. Have you tested updates from this morning?

**Comment by Queue-Fair** [comment]
I'm sorry, I don't know how to test updates from this morning.

We did notice that the above change prevented disabling leaveamessage when no operators are logged in (we don't want the button with the envelope to show when there is no-one logged in) - but we fixed that by applying the same change to the immediately preceding "hide" function like this:

{key:"hide",value:function(){setTimeout(()=>{this.restyle("display", "none  !important")},100)}}

I think that means that the shows and hides are done in the same order, but after moving the element to its final position, hence no layout shift.  It seems to work, anyway!

Hope that helps!
